Improve foreign key integrity of dump_domain_data management command

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-14803

When testing the `dump_domain_data` command with a QA domain, I stumbled into unexpected foreign key violations which led me to see that CaseDeduplicationActionDefinition was not being included in our data dump currently. It seemed best to create a test to ensure we are aware of all foreign key relationships and whether or not they are included in the data dump.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
I tested `dump_domain_data` locally to ensure the newly added models work as expected (I assume the query would have failed if it were invalid).

Additionally, this is a tool used by developers only, so it is pretty safe to make changes here.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
I added an automated test to ensure we are aware of foreign key relationships in the context of the domain dump moving forward.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
None


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Add test to ensure foreign key integrity in domain dump

Currently there are some SQL models included in the domain dump that
reference other models via a foreign key that are not included in the
domain dump. This leads to constraint violations when attempting to
load the dumped models.

Here is the output of this test at the time it was added:

AssertionError: Not all foreign key relationships will not be included in the domain dump.

data_interfaces.CaseRuleAction foreign keys to these uncovered models:
	data_interfaces.CaseDeduplicationActionDefinition

generic_inbound.ConfigurableAPI foreign keys to these uncovered models:
	userreports.UCRExpression

generic_inbound.ConfigurableApiValidation foreign keys to these uncovered models:
	userreports.UCRExpression

scheduling.AlertEvent foreign keys to these uncovered models:
	scheduling.SMSCallbackContent

scheduling.CasePropertyTimedEvent foreign keys to these uncovered models:
	scheduling.SMSCallbackContent

scheduling.RandomTimedEvent foreign keys to these uncovered models:
	scheduling.SMSCallbackContent

scheduling.TimedEvent foreign keys to these uncovered models:
	scheduling.SMSCallbackContent

users.RolePermission foreign keys to these uncovered models:
	users.Permission

gherceg committed:
Make foreign key test for domain dump pass

CaseDeduplicationActionDefinition and UCRExpression were simply in
the UNKNOWN_MODELS list, so moving those into the dumped models was
straightforward. I double checked that the filters specified for each
model are correct, and tested by running dump_domain_data locally.

SMSCallbackContent appears to no longer be in use, and there is not
an easy way to fetch it by domain anyway. It is easiest to just
confirm if a domain has any entries for this object independently when
performing a domain dump.

Permission is an interesting one. While permissions are created on
fresh install, it seems precarious to rely on the creation to remain
consistent to the point where RolePermissions will always foreign
key to the correct Permission.

gherceg commented on corehq/apps/dump_reload/tests/test_dump_models.py:
```diff
@@ -195,3 +194,34 @@ def _ignore_model(model):
+
+def test_foreign_keys_for_dumped_sql_models_are_also_included():
+    IGNORE_FK_MODELS = [
+        "scheduling.SMSCallbackContent",  # Model is no longer supported, so this would only impact old domains
```
Instead of ignoring `SMSCallbackContent`, I could add an entry in `APP_LABELS_WITH_FILTER_KWARGS_TO_DUMP` like so:
```
FilteredModelIteratorBuilder('scheduling.SMSCallbackContent', ManyFilters('alertevent__schedule__domain', 'casepropertytimedevent__schedule__domain', 'randomtimedevent__schedule__domain', 'timedevent__schedule__domain')),
```
But it didn't seem worth it since this model is likely not going to be in use anyway.

gherceg replied:
I'm second guessing ignoring `SMSCallbackContent`. The reasoning is that it seems harder to maintain documentation/awareness around this list so future operators of the `dump_domain_data` command know to confirm the domain does not have any `SMSCallbackContent` objects than it does to just include this filter in the list of dumped models. The command already takes an obscene amount of time to run, so omitting this for that reason isn't sufficient.

TLDR; `users.Permission` is a unique exception that justifies being in `IGNORE_FK_MODELS` where `SMSCallbackContent` really isn't. What do you think @kaapstorm and @millerdev?

millerdev replied:
Seems reasonable to include `SMSCallbackContent` if it will not add significant time to the dump.

gherceg replied:
Yep I tested it with a few domains in production and it is very fast. I think I phrased what I attempted to say above a bit better in the commit message: 98361a36494daa3144088ec26036aca3b8647a05.

kaapstorm commented on corehq/apps/dump_reload/tests/test_dump_models.py:
```diff
@@ -195,3 +196,24 @@ def _ignore_model(model):
+
+    assert (
+        not uncovered_fks_by_model
+    ), "Not all foreign key relationships will not be included in the domain dump.\n\n" + "\n\n".join(
```
Nit: I think you mean "Not all ... will be ..."

```suggestion
    ), "Not all foreign key relationships will be included in the domain dump.\n\n" + "\n\n".join(
```

gherceg replied:
I remember seeing this issue, but I clearly did not fix it 😅 thanks for catching!

gherceg committed:
Fix assertion text typo

Co-authored-by: Norman Hooper <708421+kaapstorm@users.noreply.github.com>

millerdev reviewed (approved):
The new test is great!

gherceg committed:
Remove SMSCallbackContent from IGNORE_FK_MODELS

    It's easier to include SMSCallbackContent in the dumped data than it is
    to ensure future operators of dump_domain_data are aware of potentially
    ignored models that impact a domain. IGNORE_FK_MODELS should be reserved
    for truly unique instances, like the users.Permission model which will
    be recreated on the fresh environment domain data is loaded into.
